src/main/scala/org/dia/algorithms/mcc/MCCOps.scala

```diff
@@ -104,72 +105,86 @@ object MCCOps {
     val partitionIndex: Int = partition._1
     val partitionStartNodeIndex: Int = partitionMaxSize * (partitionIndex-1)
     val partitionEndNodeIndex: Int = partitionMaxSize * partitionIndex
+    val  edgeList = partition._2
+    val subgraphList = new mutable.MutableList[mutable.HashSet[String]]
 
-    val firstEdge = partition._2.toSeq(0)
+    val firstEdge = edgeList.toSeq(0)
     val partitionStartFrameNum =
       if (firstEdge.metadata("index").toInt == partitionStartNodeIndex) firstEdge.srcNode.frameNum else -1
 
-    val lastEdge = partition._2.toSeq(partition._2.size - 1)
+    val lastEdge = edgeList.toSeq(edgeList.size - 1)
     val partitionEndFrameNum =
       if (lastEdge.metadata("index").toInt == partitionEndNodeIndex) lastEdge.srcNode.frameNum else -1
 
     /** To have a key and a set of values, we use MultiMap */
     val edgeMap = new mutable.HashMap[String, MCCEdge]
+    /** A map of all edges originating from the key (i.e the source node) */
+    val srcNodeMap = new mutable.HashMap[String, mutable.Set[String]] with mutable.MultiMap[String, String]
+    /** A map of edges originating or ending in the current node (key of the map) */
     val edgeMapString = new mutable.HashMap[String, mutable.Set[String]] with mutable.MultiMap[String, String]
-    val nodeMap = new mutable.HashMap[String, MCCNode]()
-    /** A set of nodes who have 0 incoming edges */
-    val originNodes = mutable.HashSet[MCCNode]()
 
-    for(edge <- partition._2) {
+
+    for(edge <- edgeList) {
       val srcKey = edge.srcNode.hashKey()
       val destKey = edge.destNode.hashKey()
       edgeMap(edge.hashKey()) = edge
+      srcNodeMap.addBinding(srcKey, edge.hashKey())
       edgeMapString.addBinding(srcKey, edge.hashKey())
+      edgeMapString.addBinding(destKey, edge.hashKey())
     }
 
     /** Edges we need to carry forward to the next iteration */
     val filteredEdges = new mutable.HashSet[String]()
-    val subgraphList = new mutable.HashSet[String]()
+
     val discardedEdges = new mutable.HashSet[String]()
+    val visitedNodes = new mutable.HashSet[String]()
 
-    for(node <- edgeMapString.keys) {
-      val graphInfo = getGraphInfo(node, 0, false, edgeMapString, new mutable.HashSet[String](),
-        partitionEndFrameNum.toString, partitionStartFrameNum.toString)
+    for(node <- srcNodeMap.keys) {
+      if (!visitedNodes.contains(node)) {
+        val graphInfo = getGraphInfo(node, 0, false, srcNodeMap, new mutable.HashSet[String](),
```

Comment:
Use tuple syntax to unbox the individual tuple elements as variables.
For example,

val (x, y, z) = getGraphInfo(...